### ISSUE TITLE
twisted.conch.ssh.agent is missing some request and response types

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8160]
|<img alt="ori_livneh's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| ori_livneh reported|
|-|-|
|Trac ID|trac#8160|
|Type|defect|
|Created|2015-12-31 09:48:23Z|

The OpenSSH's ssh-agent.
Implements the SSH v2 key agent protocol.  This protocol is documented in the

The SSH key agent protocol specification declares some request and response types that are not included in the ```messages``` dict in file scope in ```twisted.conch.ssh.agent```.

This is problem, because ```SSHAgentServer.dataReceived()``` automatically rejects any request type that is not in the messages dict. This means that anyone looking to extend SSHAgentServer so that it supports additional request types must choose between two options, both unsavory:

1. Patch in the missing request types into ```messages```.
2. Override the ```dataReceived``` method in its entirety, duplicating a dozen lines of tricky bit-twiddling code.

Apart from declaring the additional types, ```SSHAgentServer.dataReceived()``` must also be very slightly modified to preserve backward-compatibility. The code currently assumes that every type declared in the messages dict has a corresponding agentc_* method. With the introduction of the missing types, this would no longer be true, causing an AttributeError exception. Thus the code must be modified to check that the request type exists in messages _and_ that a handler method for the request type exists.

Attachments:

* [ticket_8160_ssh_agent_protocol_types.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/41a/41a73f435222075b21667758daf2fdd658cc8251/e8cf34d36144351aeead56a8b9380772e19ee73e.patch) (1683 bytes) - added by ori_livneh on 2015-12-31 09:50:26Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__8160 8160
type__defect defect
reporter__ori_livneh ori_livneh
priority__normal normal
milestone__None None
branch__ 
branch_author__ 
status__new new
resolution__None None
component__conch conch
keywords__None None
time__1451555303819706 1451555303819706
changetime__1451558261076846 1451558261076846
version__None None
owner__ori_livneh ori_livneh
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to ori_livneh|
|-|-|

Many thanks for your contribution.

------

There is a problem with the new constants defined in this patch.

They are just defined there, without any comments or documentation and without being used in the Twisted code and without any tests.

At a later time, a developer will read the code, will have a WTF moment regarding why these unused values are there and might remove them.

I don't think that we should ask developer to always check the commit history, identify the associated ticket and then read the ticket description or all comments from this ticket.

------

The changes from SSHAgentServer.dataReceived() need a test which fails on trunk but pass with current changes. Without this there is a risk of having this code refactored later and have a regression.

But I am not sure if your current changes are OK. I am -1 on them.

I think that the person who designed this code has made this behaviour intentional.

Now I now that this might be old code and by that time there were different standard, but if you are changing the 'design' of this code please document the new design.

Please see my next question.

----

My main question for this is:

Why do you want to extend SSHAgentServer ?

1. Do you want it to extend and implement standard functionalities?
2. Do you want to extend it to implement custom functionalities?

I think that for the standard functionalities it would be nice to have them contributed to Twisted and have agentc_* methods to handle the standard mesasges.

For non-standard, custom functionalities, I think that we should define an API in which extensions can register their custom message codes and custom message handlers.

----------

Every patch/contibution needs to have an associated NEWS (release notes) file fragment 

More details here https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles

In this way, your changes will be advertised in the release.

-----

In summary:

* Please define if you want standard or custom extensions.

* If you don't want to contribute your standard extensions to Twisted, please look for an API which can be used to extend the current code and which can be tested in Twisted.

* If you want to implement custom extensions, please define a public API for this.

* All changes need tests and documentation.

Please review my comments and submit a new patch.


Again, many thanks for your contribution.
